src/translator/cache.cpp

```diff
@@ -0,0 +1,163 @@
+#include "cache.h"
+
+#include <cstdlib>
+
+#include "translation_model.h"
+
+namespace marian {
+namespace bergamot {
+
+size_t TranslationCache::hash(const TranslationCache::CacheKey &key) {
+  size_t seed = 42;
+  for (auto &word : key.words) {
+    size_t hashWord = static_cast<size_t>(word.toWordIndex());
+    util::hash_combine<size_t>(seed, hashWord);
+  }
+
+  util::hash_combine<size_t>(seed, key.model.modelId());
+  return seed;
+}
+
+ThreadSafeL4Cache::ThreadSafeL4Cache(const ThreadSafeL4Cache::Config &config)
+    : epochManagerConfig_(config.ebrQueueSize, std::chrono::milliseconds(config.ebrIntervalInMilliseconds),
+                          config.ebrNumQueues),
+      cacheConfig_(config.sizeInMB * 1024 * 1024, std::chrono::seconds(config.timeToLiveInMilliseconds),
+                   config.removeExpired),
+      service_(epochManagerConfig_),
+      context_(service_.GetContext()),
```

Comment:
I don't understand the lifecycle management of the L4 values you're reading enough.  Currently you have one read `context_` that is used for all read operations across all threads.  

Supposedly `context_` keeps things alive:

https://github.com/Microsoft/L4/wiki/Epoch-Queue

Does that mean reads are just accumulating?  What is decrementing the reference count?

Reply:
> Does that mean reads are just accumulating? What is decrementing the reference count?

This is correct. They are accumulating. Fixes have been pushed, however, L4 is not lock-free on read-anymore in this case.

Reply:
@XapaJIaMnu 
1. Buckets matter now. [test_cache_hparam.sh](https://github.com/browsermt/bergamot-translator-tests/blob/cache-rt/cache-tests/test_cache_hparam.sh) http://ix.io/3AUP 
2. The multi/single (40/1) cache (on/off) 1M/100K experiment [test_cache_overhead.sh](https://github.com/browsermt/bergamot-translator-tests/blob/cache-rt/cache-tests/test_cache_overhead.sh) http://ix.io/3AUQ

Reply:
I'm confused, were the previous experiments 1thread?

Reply:
> were the previous experiments 1thread?

Writes weren't flushed until destruction of `Service` -> `Cache` -> `L4:`, so not a proper contention setting. L4 doesn't allow lock-free reads anymore, limitation of API @ `context_`. 

L4 allows a second reader to accept context and read "multiple" times without locks as demonstrated with multiple key-value pairs in https://github.com/browsermt/L4/blob/master/Examples/main.cpp, while something else writes I suppose. For other purposes, our code is no-longer "lock-free" (be it read or write).

Reply:
I have missed this conversation. Why is the code no-longer lock-free on read? Was what was done before wrong?